cpp/src/groupby/sort/group_merge_m2.cu

```diff
@@ -14,154 +14,90 @@
 struct merge_fn {
-  size_type const* const d_offsets;
-  size_type const* const d_counts;
-  result_type const* const d_means;
-  result_type const* const d_M2s;
+  size_type const* d_offsets;
+  count_type const* d_counts;
+  result_type const* d_means;
+  result_type const* d_M2s;
 
-  auto __device__ operator()(size_type const group_idx) noexcept
+  auto __device__ operator()(size_type const group_idx) const
   {
-    auto const start_idx = d_offsets[group_idx], end_idx = d_offsets[group_idx + 1];
-
-    // This case should never happen, because all groups are non-empty as the results of
-    // aggregation. Here we just to make sure we cover this case.
-    if (start_idx == end_idx) {
-      return thrust::make_tuple(size_type{0}, result_type{0}, result_type{0}, int8_t{0});
-    }
+    count_type n{0};
+    result_type avg{0};
+    result_type m2{0};
 
-    // If `(n = d_counts[idx]) > 0` then `d_means[idx] != null` and `d_M2s[idx] != null`.
-    // Otherwise (`n == 0`), these value (mean and M2) will always be nulls.
-    // In such cases, reading `mean` and `M2` from memory will return garbage values.
-    // By setting these values to zero when `n == 0`, we can safely merge the all-zero tuple without
-    // affecting the final result.
-    auto get_partial_result = [&] __device__(size_type idx) {
-      {
-        auto const n = d_counts[idx];
-        return n > 0 ? partial_result<result_type>{n, d_means[idx], d_M2s[idx]}
-                     : partial_result<result_type>{size_type{0}, result_type{0}, result_type{0}};
-      };
-    };
-
-    // Firstly, store tuple(count, mean, M2) of the first partial result in an accumulator.
-    auto accumulator = accumulate_fn<result_type>{get_partial_result(start_idx)};
-
-    // Then, accumulate (merge) the remaining partial results into that accumulator.
-    for (auto idx = start_idx + 1; idx < end_idx; ++idx) {
-      accumulator(get_partial_result(idx));
+    auto const start_idx = d_offsets[group_idx], end_idx = d_offsets[group_idx + 1];
+    for (auto idx = start_idx; idx < end_idx; ++idx) {
+      auto const partial_n = d_counts[idx];
+      if (partial_n == 0) { continue; }
+      auto const partial_avg = d_means[idx];
+      auto const partial_m2  = d_M2s[idx];
+      auto const new_n       = n + partial_n;
+      auto const delta       = partial_avg - avg;
+      m2 += partial_m2 + delta * delta * n * partial_n / new_n;
+      avg = (avg * n + partial_avg * partial_n) / new_n;
+      n   = new_n;
     }
 
-    // Get the final result after merging.
-    auto const& merge_vals = accumulator.merge_vals;
-
     // If there are all nulls in the partial results (i.e., sum of all valid counts is
     // zero), then the output is a null.
-    auto const is_valid = int8_t{merge_vals.count > 0};
-
-    return thrust::make_tuple(merge_vals.count, merge_vals.mean, merge_vals.M2, is_valid);
+    auto const is_valid = n > 0;
+    return thrust::make_tuple(n, avg, m2, is_valid);
```

Comment:
CCCL is deprecating `thrust::tuple` and its utlities. Let's use `cuda::std::tuple` instead.
```suggestion
    return cuda::std::tuple{n, avg, m2, is_valid};
```

Reply:
👍 Is this the deprecation you're mentioning?
https://github.com/NVIDIA/cccl/issues/4305

Reply:
I don't have the exact PR number in mind but 4305 is definitely related.

Reply:
Unfortunately this doesn't work yet,, since `thrust::make_zip_iterator` still internally uses `thrust::tuple`:
```
template <typename... Iterators>
_CCCL_HOST_DEVICE zip_iterator<thrust::tuple<Iterators...>> make_zip_iterator(thrust::tuple<Iterators...> t)
{
  return zip_iterator<thrust::tuple<Iterators...>>(t);
} // end make_zip_iterator()

template <typename... Iterators>
_CCCL_HOST_DEVICE zip_iterator<thrust::tuple<Iterators...>> make_zip_iterator(Iterators... its)
{
  return make_zip_iterator(thrust::make_tuple(its...));
} // end make_zip_iterator()
```